Add policy section to docs

tcharding opened the pull request:
In an effort to consolidate knowledge spread out over time in various places on GitHub add a `Policy` section to `CONTRIBUTING.md`.

Add initial sections on import statements, errors, rustdocs,attributes, and licensing.

tcharding commented:
Possible other candidate for inclusion, info on SPDX e.g.,

Files should start with attribution and license information as follows:
```rust
// Bitcoin - Written by the rust-bitcoin developers.
// SPDX-License-Identifier: CC0-1.0
```

apoelstra commented:
Neat! A good start. I'm sure others will chime in with things they think we ought to add.

We should require three ACKs for this PR.

apoelstra reviewed (dismissed):
ACK a954f0432a1dc63e71d7eb5018da39fb62c5ce9b

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -204,6 +205,56 @@ library maintainers on the exclusion from this rule. In such cases there is a
+#### Errors
+
+Return as much context as possible with errors e.g., if an error was encountered
+parsing a string include the string in the returned error type.
```
Error types should be `non_exhaustive` and have private fields unless you're very confident they won't change.
Error types should not commit to implementing traits they may not be able to implement int the future, especially `Copy` but possibly also `Eq` and `Ord`. They should however commit to `Send` and `Sync`.
If the error contains source implement `Display` using `write_err!()` macro.
If a function consumes costly-to-compute input (allocations are also considered costly) it should return it back in the error type.
Errors should have `Error` suffix and conditionally implement `std::error::Error` if they are public.

tcharding replied:
Moving this thread over to the error epic

tcharding replied:
Should we have a small set of derives that are used on all error types? `Clone, Debug, PartialEq`?

tcharding replied:
> They should however commit to `Send` and `Sync`

I don't know what action is required to achieve this, I imagine other readers of the doc may not either. Can you explain further please?

Kixunil replied:
`Debug` is the only one I'd guarantee. `Clone` is very likely for most error types but there could be exception when we want to return a resource instead of consuming it. (OTOH `&mut` is probably better)

> I don't know what action is required to achieve this

Not putting things like `Cell` or `Rc` into `Error` types. Sadly it could happen by accident without the compiler complaining and we may break the user code. Maybe we should have something like this:

```rust
// #[allow(unused)] may be required
trait AssertSendSync: Send + Sync {
}

// Perhaps we could also check other traits
macro_rules! check_pub_error {
    ($type:ty) => {
        impl AssertSendSync for $ty {}
    }
}
```

tcharding replied:
Done! https://github.com/rust-bitcoin/rust-bitcoin/pull/1148

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -204,6 +205,56 @@ library maintainers on the exclusion from this rule. In such cases there is a
+
+#### rustdocs
+
+Be liberal with references to BIPS, the aim is that devs can learn about Bitcoin
```
BIPs or other documentation

Kixunil reviewed (commented):
Concept ACK, few suggestions.

tcharding commented:
> We should require three ACKs for this PR.

If/when bors comes about we should consider if there is some way to flag PRs that require more acks.

Kixunil commented:
I'd like some rules that require more acks based on files changed or so, if we have any critical code or policies. Might be reasonably easy to add.

tcharding commented:
Canonical license blurb, open under discussion still: https://github.com/rust-bitcoin/rust-bitcoin/pull/1138

tcharding commented:
Putting on ice until after release of 0.30.0

tcharding commented:
> I'd like some rules that require more acks based on files changed or so, if we have any critical code or policies. Might be reasonably easy to add.

Just flagging that I took this PR off draft but have not actioned this comment of yours @Kixunil, not sure what to do about it?

apoelstra commented:
Our current policy is roughly "if one of the 2 ACKers says that they want a 3rd ACK, then we do it". I wouldn't mind codifying this. But I can't think of a simple mechanical rule that would say "you need 3 ACKs if you touch this file" or something like that.

apoelstra commented on CONTRIBUTING.md:
```diff
@@ -189,6 +190,81 @@ library maintainers on the exclusion from this rule. In such cases there is a
+    /// Documention for variant B.
+    B,
+}
+pub_errro_type!(Error)
```
In 8578505bf4c2f3f01116c3a97715e2bfd1f88c64: typo

apoelstra commented on CONTRIBUTING.md:
```diff
@@ -189,6 +190,81 @@ library maintainers on the exclusion from this rule. In such cases there is a
+
+#### rustdocs
+
+Be liberal with references to BIPS or other documention, the aim is that devs can learn about
```
In 8578505bf4c2f3f01116c3a97715e2bfd1f88c64: run-on (replace comma with semicolon or period)

tcharding replied:
Fixed, and the 'documention' typo.

apoelstra commented on CONTRIBUTING.md:
````diff
@@ -189,6 +190,81 @@ library maintainers on the exclusion from this rule. In such cases there is a
+    pub version: i32,
+```
+
+Do use rustdoc subheadings. Do put an empty newline below each heading e.g.,
````
In 8578505bf4c2f3f01116c3a97715e2bfd1f88c64:

Maybe we should also say that you are required to add a `Panics` section if the function panics under any circumstances (including array indexing on input). For `Errors` it's probably ok to let this be a judgement call since some functions have "obvious" error returns.

tcharding replied:
Oh and we could mention `track_caller`  (your mention of array index out of bound triggered this thought).

tcharding commented:
changes in force push:
- added "Attributes" section
- fixed typos
- Added to `#[non_exhaustive]` in error:  `// Add liberally; if the error type may ever have new variants added.`

sanket1729 reviewed (dismissed):
ACK 4d630e0ac386eb26a975429af590ea52219362b0

tcharding commented:
Soz @sanket1729, the rebase is just to remove leading patch that was from another PR. No other changes.

sanket1729 reviewed (approved):
ACK 271338d1036987d671fc9a6c517dc853d4a297ac

apoelstra commented:
Will wait for @Kixunil to ACK on this before merging.

apoelstra reviewed (dismissed):
ACK 271338d1036987d671fc9a6c517dc853d4a297ac

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -189,6 +190,101 @@ library maintainers on the exclusion from this rule. In such cases there is a
+#### Errors
+
+Return as much context as possible with errors e.g., if an error was encountered parsing a string
+include the string in the returned error type. If a function consumes costly-to-compute input
```
This one I'm no longer sure about. Didn't have time to investigate deeply yet.

tcharding replied:
Flagging, I have not changed these lines as yet.

Kixunil replied:
I now again believe it's the right thing to do

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -189,6 +190,101 @@ library maintainers on the exclusion from this rule. In such cases there is a
+
+- be `non_exhaustive`
+- have private fields unless you're very confident they won't change.
+- only derive  `Debug`
```
Might be useful to put emphasis on "derive" so that people don't confuse it with "implement"

tcharding replied:
Done.

Kixunil commented on CONTRIBUTING.md:
````diff
@@ -189,6 +190,101 @@ library maintainers on the exclusion from this rule. In such cases there is a
+```rust
+pub struct BlockHeader {
+    /// Originally protocol version, but repurposed for soft-fork signaling.
+    ///
````
Also possibly `<details>` section for Bitcoin beginners.

tcharding replied:
Nice, added.

Kixunil commented on CONTRIBUTING.md:
````diff
@@ -189,6 +190,101 @@ library maintainers on the exclusion from this rule. In such cases there is a
+    pub fn iter_funding_utxos(&self) -> impl Iterator<Item = Result<&TxOut, Error>> {
+```
+
+Add `Panics` section if the function may panic under any circumstance (including array index out of bounds)
````
I think it should instead say:

> Add `Panics` section if any *input* to the function can trigger a panic.
> Generally we prefer to have non-panicking API but it is impractical in some cases.
> If you're not sure, feel free to ask. If we determine panicking is more practical it must be documented.
> Internal panics that could be theoretically occur because of bugs in our code must **not** be documented.

tcharding replied:
Used as suggested (with minor grammar/typo fix)

Kixunil commented on CONTRIBUTING.md:
````diff
@@ -189,6 +190,101 @@ library maintainers on the exclusion from this rule. In such cases there is a
+```
+
+
+#### rustdocs
````
Capital R

tcharding replied:
Done

Kixunil commented on CONTRIBUTING.md:
````diff
@@ -189,6 +190,101 @@ library maintainers on the exclusion from this rule. In such cases there is a
+impl std::error::Error for ParseAmountError {
+```
+
+- `#[track_caller]`: Used any time a function may panic and the stack trace is uninformative.
````
More like "used on functions that panic on invalid arguments"

tcharding replied:
Cool, done.

Kixunil reviewed (commented):
I' not sure if we should merge it if there's a chance we'll change it soon.

tcharding commented:
Changed to use a dummy `FooBar` type for rustdocs example.

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -189,6 +190,117 @@ library maintainers on the exclusion from this rule. In such cases there is a
+  (see https://rustc-dev-guide.rust-lang.org/backend/implicit-caller-location.html)
+
+- `#[cfg(rust_v_1_60)]`: Used to guard code that should only be built in if the toolchain is compatible.
+  These configuration conditionals are set at build time in `bitcoin/build.rs`.
```
Possibly mention "new version attributes may be added as needed"

tcharding replied:
Added, thanks.

Kixunil commented:
Regarding parsing errors, I intend to write it down and open a discussion in URLO when I get a bit more time.

tcharding commented:
I don't know what the status of this PR is, do we want it:
1. merged?
2. closed?
3. left open to be re-visited at a later stage?

Kixunil commented:
Leave it open, hopefully I write it up during weekend.

tcharding commented:
No sweat, take your time.

Kixunil commented:
Here's the discussion: https://users.rust-lang.org/t/error-design-to-store-the-context-or-to-not/91815

tcharding commented:
Converting to draft to indicate not-mergable.

tcharding commented:
Note to self: Add to rustdoc section something about putting important info in the first few words of rustdoc heading.

https://github.com/rust-bitcoin/rust-bitcoin/issues/1920#issuecomment-1638173124

tcharding commented:
Force push just added import style policy, I'm still to see to review suggestions.

apoelstra commented:
903b2074f8d3117dcd03b4f48a92726d7db0f272 looks good to me except:

* I think the `Errors` section is out of date with what we want (do we still want to "only derive `Debug`?) but I'm not sure.
* The docrs attribute can now be removed :)

tcharding commented:
Back to draft because this depends on https://github.com/rust-bitcoin/rust-bitcoin/pull/1743

tcharding commented:
Just removed the mention of `check_pub_error` instead.

apoelstra reviewed (approved):
ACK 539b8ad2b9af34ed33113d6738d6e3779dfa7b5a

Kixunil commented on CONTRIBUTING.md:
```diff
@@ -228,6 +229,137 @@ library maintainers on the exclusion from this rule. In such cases there is a
+#[rustfmt::skip]                // Keep public re-exports separate.
+pub use {
+    crate::aa_aa_this;
+    crate::bb_bb::that::{Other};
```
This looks weird :)

tcharding replied:
Dunno what I was doing when I wrote this, the `;`s are wrong too. Fixed and force pushed, thanks.

apoelstra reviewed (dismissed):
ACK fe1e8aef8cd24b9695d0effe7fc442a2b8b4ddd0

sanket1729 reviewed (dismissed):
ACK fe1e8aef8cd24b9695d0effe7fc442a2b8b4ddd0

apoelstra commented:
Since @Kixunil is around I'll let him ACK as well before merging.

tcharding commented:
On ice until next release.

yancyribbens commented:
Is it worth adding mention of how silent overflows should be handled (or not handled)? https://github.com/rust-bitcoin/rust-bitcoin/issues/2086

apoelstra commented:
I like the idea of having a project-wide overflow policy. But I have no idea what such a policy should be :). Feerates have different rules from amounts, which have different rules from sizes, etc.

apoelstra reviewed (approved):
ACK 3bebecc7ea7a775b097c714425cf56009ef0b3eb

tcharding commented:
Can this go in?

apoelstra commented:
cc @sanket1729 @Kixunil we need a second ack here

Kixunil reviewed (approved):
ACK 3bebecc7ea7a775b097c714425cf56009ef0b3eb
